### PR TITLE
Allow shared services to be annotated

### PR DESCRIPTION
**Pull request by abstratt**
This could make usesService() optional/unnecessary.

WIP

Issue: #16170

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/ResourceLock.java [hunk 1]
```diff
@@ -23,6 +23,14 @@
  * outside of a {@link ResourceLockCoordinationService#withStateLock(Transformer)} transform.
  */
 public interface ResourceLock extends Describable {
+    enum Kind {
```
Comment (bamboo) [inline subprojects/base-services/src/main/java/org/gradle/internal/resources/ResourceLock.java:26]: ❓ How is this related?

File: subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy [hunk 2]
```diff
@@ -71,33 +71,62 @@ class BuildServiceIntegrationTest extends AbstractIntegrationSpec {
...
+
+        then:
+        outputDoesNotContain "'Task#usesService'"
     }
 
-    def "service is created once per build on first use and stopped at the end of the build"() {
+    def "does not nag when service is annotated with @ServiceReference and feature preview is enabled"() {
+        given:
         serviceImplementation()
+        customTaskUsingServiceViaProperty("@${ServiceReference.class.name}(name = 'foobar')")
```
Comment (bamboo) [inline subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy:104]: ❓ Wouldn't `name = 'foobar'` have to match a registered service name?

File: subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java [hunk 3]
```diff
@@ -73,6 +76,15 @@ public BuildServiceProvider(
...
         this.internalServices = internalServices;
         this.listener = listener;
+        this.maxUsages = maxUsages;
+    }
+
+    public int getMaxUsages() {
+        return maxUsages;
+    }
+
+    public boolean isLockRequired() {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java:86]: ❓ What's going on here?

Comment (abstratt) [reply subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java:86]: This is to allow the nagger to determine whether a service reference requires locking but is not part of the list of resources locked by the task. Not all service references require locks (for instance, locks are not created for services that do not impose maximum parallel usage).

Before, we required every service to be declared in the list of required services, so just checking that list in the nagger [was enough](https://github.com/gradle/gradle/blob/0dad3eab937f2b0f3bea0daaf42ef13c828c401e/subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java#L46).  As things were implemented in this spike, that list holds only the list of service names explicitly declared as used (either using #usesService()  or using the name in the `@ServiceReference` annotation), but does not include the list of services injected manually. The actual list we care about is the list of locks, as checked [here](https://github.com/gradle/gradle/blob/0dad3eab937f2b0f3bea0daaf42ef13c828c401e/subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java#L37).

File: subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy (hunk 2 above)
Comment (abstratt) [reply subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy:104]: Note that this PR, although declares a `name` property on the ServiceReference annotation, does not care about the value provided. 

This other PR implemented that:  https://github.com/gradle/gradle/pull/21754/files#diff-177b01df4f530a32fd73f07c48f1767fb6ce34465477be561e6a71f163434f36R104

File: subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java (hunk 3 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java:86]: I see, thanks. For the final implementation it would be better not to conflate these things, i.e., service reference properties and changing how locks are checked, if possible.
